src/irmin/contents.ml

```diff
@@ -16,6 +16,85 @@
 
 open Lwt.Infix
 
+type json = [
+  | `Null
+  | `Bool of bool
+  | `String of string
+  | `Float of float
+  | `O of (string * json) list
+  | `A of json list
+]
+
+let json =
+  let open Type in
+  mu (fun ty ->
+  variant "json" (fun null bool string float obj arr -> function
+    | `Null -> null
+    | `Bool b -> bool b
+    | `String s -> string s
+    | `Float f -> float f
+    | `O o -> obj o
+    | `A a -> arr a)
+  |~ case0 "null" `Null
+  |~ case1 "bool" bool (fun x -> `Bool x)
+  |~ case1 "string" string (fun x -> `String x)
+  |~ case1 "float" float (fun x -> `Float x)
+  |~ case1 "object" (list (pair string ty)) (fun obj -> `O obj)
+  |~ case1 "array" (list ty) (fun arr -> `A arr)
+  |> sealv)
+
+let assoc k l =
+  try Some (List.assoc k l) with Not_found -> None
+
+let merge_objects j a b =
+  let equal = Type.equal json in
+  try
+    let v = List.fold_right (fun (k, v) acc ->
+      match assoc k a, assoc k b with
+      | Some x, Some y when not (equal x y) -> failwith "Unable to merge JSON objects"
+      | Some x, _ -> (k, x) :: acc
+      | None, Some x -> (k, x) :: acc
+      | None, None -> (k, v) :: acc
+    ) j []
+    in Merge.ok (`O v)
+  with
+    Failure msg -> Merge.conflict "%s" msg
+
+let merge2 a b =
+  match a, b with
+  | `Null, a | a, `Null -> Merge.ok a
+  | `Float a, `Float b when a <> b -> Merge.conflict "Conflicting float values"
+  | `String a, `String b when not (String.equal a b) -> Merge.conflict "Conflicting string values"
+  | `Bool a, `Bool b when a <> b -> Merge.conflict "Conflicting bool values"
+  | `O a, `O b -> merge_objects [] a b
+  | `A a, `A b ->
+      if List.for_all2 (Type.equal json) a b then
+        Merge.ok (`A a)
+      else
+        Merge.conflict "Unable to merge JSON arrays"
+  | `Float _, `Float _ | `String _, `String _ | `Bool _, `Bool _ -> Merge.ok a
+  | _, _ -> Merge.conflict "Conflicting JSON values"
+
+let merge3 a b c =
+  let open Merge.Infix in
+  match a, b, c with
+  | `Null, a, b -> merge2 a b
+  | `O a, `O b, `O c -> merge_objects a b c
+  | a, b, c -> (merge2 a b >>=* fun b -> merge2 b c)
+
+let merge_json' ~old a b =
+  let equal = Type.equal json in
+  let open Merge.Infix in
+  old () >>=* function
+    | Some old ->
+        if equal old a then Merge.ok b
+        else if equal old b then Merge.ok a
+        else merge3 old a b
+    | None ->
+        merge2 a b
```

Comment:
the proper way to write these merge functions is to define mutually recursive functions which match the shape of `json`: for instance, you want the recursive merge function to be applied on sub-trees too. I think the result should be very similar to what it is done in [Tree.merge](https://github.com/mirage/irmin/blob/master/src/irmin/tree.ml#L908) and [Tree.Node.merge](https://github.com/mirage/irmin/blob/master/src/irmin/tree.ml#L440-L460) but with a different (and somehow a bit simpler, as their no indirection) datatype.

Reply:
Okay, thank you for the tip. I will stop trying to make things too simple!